src/lib/shadow-shim-helper-rs/src/rootedcell/mod.rs

```diff
@@ -104,3 +107,74 @@ impl Default for Root {
         Self::new()
     }
 }
+
+pub struct RootScope<T> {
+    // This doesn't *really* need to behave as a T, but in intended usage T
+    // shouldn't introduce any constraints. It's just a unique marker type.
+    _phantom: PhantomData<T>,
+}
+
+impl<T> RootScope<T> {
+    thread_local!(static CURRENT: std::cell::RefCell<Option<RefCell<Root>>>  = std::cell::RefCell::new(None));
+
+    pub fn with_current_set_to(root: Root, f: impl FnOnce()) -> Root {
+        Self::CURRENT.with(|current| {
+            let prev = current.replace(Some(RefCell::new(root)));
+            f();
+            // Will panic if there are any live borrows.
+            // Note the `as_mut` to avoid actually moving the inner RefCell before `take`
+            // has done that validation.
+            let root = current.borrow_mut().as_mut().unwrap().take();
+            current.replace(prev);
+            root
+        })
+    }
+
+    pub fn current() -> Option<impl Deref<Target = Root> + 'static> {
+        Self::CURRENT.with(|current| {
+            let outer = current.borrow();
+            if let Some(root) = outer.as_ref() {
+                let ret = root.borrow();
+                type ReturnType = std::cell::Ref<'static, Root>;
+                static_assertions::assert_not_impl_all!(ReturnType: UnwindSafe);
+                // SAFETY:
+                // We only mutate the RefCell that Ref is borrowed from via `RefCell::take`,
+                // which validates at run-time that there are no borrows outstanding.
+                //
+                // We validated above that the type we are using is `!UnwindSafe`, since
+                // otherwise `catch_unwind` could potentially circumvent this check.
+                let static_ret: ReturnType = unsafe { std::mem::transmute(ret) };
+                Some(static_ret)
```

Comment:
I don't think setting `!UnwindSafe` really helps us here. For example, here's a very contrived example where we can move the `Ref` outside of the closure using `catch_unwind` and cause UB:

```rust
#[test]
fn foo() {
    thread_local! {
        static FOO: RefCell<Option<Box<dyn Deref<Target = Root> + 'static>>> = RefCell::new(None);
    }

    println!();

    let root = Root::new();
    let _ = std::panic::catch_unwind(|| {
        HostRootScope::with_current_set_to(root, || {
            FOO.with(|foo| {
                let root = HostRootScope::current().unwrap();
                foo.borrow_mut().replace(Box::new(root));
            });
            // FOO contains a Ref that points to the Root in CURRENT

            // will panic here
        });
    });

    // CURRENT still holds the Root because we panicked, so FOO still points to it
    let tag = FOO.with(|foo| {
        let foo = foo.borrow();
        let root = foo.as_ref().unwrap().deref().deref();
        root.tag()
    });

    let root = Root::new();
    HostRootScope::with_current_set_to(root, || {
        // CURRENT now holds a new Root
        FOO.with(|foo| {
            // when we read the Root from FOO, it's reading the new Root in CURRENT, which means
            // the memory was changed even though we hold an immutable reference to it
            let foo = foo.borrow();
            let root = foo.as_ref().unwrap().deref().deref();
            assert_eq!(
                tag,
                root.tag(),
                "We hold a `Ref<'_, Root>`, but the tag changed"
            );
        });
    });
}
```

Reply:
One solution could be to "poison" the `CURRENT` thread local if `with_current_set_to` panics. So add a flag in `CURRENT`, and then have a struct with a drop handler that poisons `CURRENT` by setting the flag, and if everything completes successfully, then `mem::forget` the struct.

There's still a question about if you have the `Root` in one thread local and a `Ref<'static, Root>` in another thread local, the thread local containing the `Ref` might get dropped before the `Root`. So maybe you could surround `CURRENT` with a `ManuallyDrop` so that it's never dropped unless you do it explicitly.